### PR TITLE
#244 - replace `search-index` with a handwritten `IndexLevel`

### PR DESCRIPTION
**Pull request by dcrousso**
this allows us to have more control over how the data is stored and how performant it is

ultimately we only need a small subset of `search-index` functionality, so this lets us be more streamlined too

included some super basic benchmarks

```
$ node benchmarks/store/index/search-index.js
create 40
clear - before 477
put 23954
query - equal 23
query - range 24
clear - after 437
```

```
$ node benchmarks/store/index/index-level.js
create 43
clear - before 429
put 133
query - equal 69
query - range 29
clear - after 482
```

this shows that while `IndexLevel` is slightly slower with exact equality `query` (which makes some sense as `IndexLevel.prototype.findExactMatches` actually uses `gt` because `level` doesn't have an `eq`) it is *massively* faster with `put` and basically equal for everything else

### CODE REVIEW

**Comment by csuwildcat** [comment]
@dcrousso have you confirmed with @mistermoe and @thehenrytsai that this supports the range of queries/indexing needed for the rest of the features planned before we hit v1? I ask because I know Moe has done some work in this area and Henry has knowledge of the needs that are upcoming, so I wanted to make sure our bases are covered.

**Comment by mistermoe** [comment]
@csuwildcat i wrote the handrolled search-index. it has support for: and, or, range

**Comment by csuwildcat** [comment]
@mistermoe I'd turn to @thehenrytsai for that, but if it's the one you already have talked about, that's probably the known good set of features we'd want. Henry, can you confirm?

File: src/store/index-level.ts [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    await Promise.all(promises);
+
+    const docIds: string[] = [];
+    for (const docId in matches) {
+      if (matches[docId] === requiredMatchCount) {
+        docIds.push(docId);
+      }
+    }
+    return docIds;
+  }
```
Comment (mistermoe) [inline src/store/index-level.ts:118]: i find this a bit hard to follow compared to:

```typescript
  async query(filter: Filter): Promise<Array<string>> {
    const matches: { [docId: string]: number } = {};
    const promises: Promise<void>[] = [];

    for (const propertyName in filter) {
      const propertyValue = filter[propertyName];

      if (typeof propertyValue === 'object' && propertyValue !== null) {
        if (Array.isArray(propertyValue)) {
          // OR query
          for (const value of propertyValue) {
            const promise = this.buildTermQuery(propertyName, value, matches);
            promises.push(promise);
          }
        } else {
          const promise = this.buildRangeQuery(propertyName, propertyValue, matches);
          promises.push(promise);
        }
      } else {
        const promise = this.buildTermQuery(propertyName, propertyValue, matches);
        promises.push(promise);
      }
    }

    const numFilters = Object.keys(filter).length;
    const docIds: string[] = [];
    await Promise.all(promises);

    for (const docId in matches) {
      if (matches[docId] === numFilters) {
        docIds.push(docId);
      }
    }

    return docIds;
  }
```

Comment (dcrousso) [reply src/store/index-level.ts:118]: what is it about the way i've written it that's more confusing?

IMO passing in `matches` and then suddenly having it be magically populated with things after the `await` is more confusing because i have to look in three different places to see all the ways that `matches` is modified (whereas the way i've written it is all in the same function).  TL;DR: i **_really_** dislike using out parameters in JS :(

would it be clearer if instead of having it take a `callback` it returned an array of all the matches?  thinking about this a bit more i think that'd be even more clean

File: src/store/index-level.ts [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  _id: string,
+  [property: string]: unknown
+};
+
+export type ExactFilter = string | number | boolean;
+
+export type RangeFilter = {
+  gte?: string,
+  lte?: string
+};
```
Comment (mistermoe) [inline src/store/index-level.ts:25]: @dcrousso thoughts on:

```typescript
type Gt = ({ gt: string } & { gte?: never }) |
                   ({ gt?: never } & { gte: string });
type Lt = ({ lt: string } & { lte?: never }) |
                   ({ lt?: never } & { lte: string });

export type RangeFilter = (Gt | Lt) & Partial<Gt> & Partial<Lt>;
```

* at least one of `gt`, `gte`, `lt`, or `lte` is required
* `gt` and `gte` are mutually exclusive.
* `lt` and `lte` are mutually exclusive.

has no benefit w.r.t. runtime checks because type erasure. but does provide a better developer experience

**Comment by mistermoe** [comment]
@shamilovtim if you have time, can we publish an unstable version of this and have you try and use it now that `search-index` is gone?

**Review by thehenrytsai — CHANGES_REQUESTED** [review]
I mean... oof.  Definitely not a trivial piece of code for my little brain given how small it is!

Must say I am not exactly beaming with confidence with the `index-level.ts` impl: the usage of the constant chars feels quite brittle, the occurrences of `__${_id}__keyPrefixes` and splitting by `@` feels a bit hacky/janky, but I guess if it works, it works... is this how `search-index` impl feels like and far worse? LOL!

On the plus side, the file size is small, let's you index boolean type, support basic `or` filter and lets me do `lt`.

This is all my brain has for today, might revisit to see if I can add more comments tomorrow morning when I mind is more fresh.

**Comment by thehenrytsai** [comment]
@csuwildcat:
> Henry, can you confirm?

This impl certainly ticks all the boxes that are needed by the latest/current version of the DWN SDK. It also adds basic `or` support at individual property level, though its usefulness is yet to be seen as we don't have a scenario/use for it yet.

The two features that are missing looking ahead (though to be clear, also missing in `search-index`) that comes to mind right now are:

1. Multi-value/array indexes. This is needed for the seemingly trivial (but not) query of "get me records that are attested by alice", where, `attesters` is a multi-value property.
1. Sorting, up until now sorting is done entirely on the client-side in memory which is hacky. This is particularly needed when dataset it large and we introduce paging. The way the new index lib is implemented seems like this is still tricky.

**Comment by mistermoe** [comment]
@thehenrytsai re: jankiness 

there's only so much one can do with a flat k/v store! 

i have benchmark files if you're interested in running them. the hand-rolled one is magnitudes faster. 

additionally:
* `search-index` made the use of dwn-sdk in react native without rolling an entirely new `MessageStore` a non-option
* `search-index` requires the most polyfills

File: src/store/index-level.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+    let requiredMatchCount = 0;
+
+    const promises: Promise<Matches>[] = [ ];
+    for (const propertyName in filter) {
+      const propertyFilter = filter[propertyName];
+
+      if (typeof propertyFilter === 'object' && propertyFilter !== null) {
+        if (Array.isArray(propertyFilter)) {
+          for (const propertyValue of propertyFilter) {
```
[the remark this answers is not in the document]
Comment (mistermoe) [reply src/store/index-level.ts:92]: good call @thehenrytsai. could `Array.from(new Set([3, 3, 3]))` before looping

File: src/interfaces/records/types.ts [hunk 4]
```diff
@@ -55,6 +55,11 @@ export type RecordsQueryFilter = {
   dateCreated?: RangeCriterion;
 };
 
+/**
+ * An exact criterion in a query filter.
+ */
+export type ExactCriterion = unknown;
```
[the remark this answers is not in the document]
Comment (dcrousso) [reply src/interfaces/records/types.ts:61]: `buildExactQueryTerms` allows for any type that it will then `flatten`

File: src/store/index-level.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+    const iteratorOptions: AbstractIteratorOptions<string, string> = {
+      gt: propertyKey
+    };
+
+    return this.findMatches(propertyKey, iteratorOptions, options);
+  }
+
+  private async findRangeMatches(propertyName: string, range: RangeFilter, options?: IndexLevelOptions): Promise<Matches> {
+    const propertyKey = this.generateKey(propertyName);
```
[the remark this answers is not in the document]
Comment (dcrousso) [reply src/store/index-level.ts:150]: it's mainly a consistency thing so that if we ever decide to adjust how `generateKey` works then it'll immediately apply everywhere instead of all places except this one

File: src/interfaces/records/types.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (dcrousso) [reply src/interfaces/records/types.ts:61]: FYI im also planning to adjust/remove this in a followup

File: src/store/index-level.ts [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+import { abortOr } from '../utils/abort.js';
+import { createLevelDatabase } from './create-level.js';
+import flat from 'flat';
+
+const KEY_DELIMITER = '<';
+const KEY_BOUNDARY = '>';
+
+const VALUE_DELIMITER = '=';
+
+const KEY_PREFIX_DELIMITER = '@';
```
[the remark this answers is not in the document]
Comment (dcrousso) [reply src/store/index-level.ts:13]: i was trying to find things that are unlikely to be used in keys or generated by base64 things or etc.

i think a better approach might actually be to leverage `JSON.stringify` to guarantee no chance of collision

File: src/store/index-level.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+export type Entry = {
+  _id: string,
+  [property: string]: unknown
+};
+
+export type ExactFilter = string | number | boolean;
+
+export type RangeFilter = {
+  gte?: string,
+  lte?: string
```
[the remark this answers is not in the document]
Comment (dcrousso) [reply src/store/index-level.ts:24]: why dont we want to allow for the "or equal" part?  it seems a bit unnecessary (especially at such a generic low-level like this) to restrict it to `lt` only

**Comment by shamilovtim** [comment]
> @shamilovtim if you have time, can we publish an unstable version of this and have you try and use it now that `search-index` is gone?

I'm ready for you guys. Go ahead and push whenever you make the latest feedback changes and I'll test

File: src/store/index-level.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mistermoe) [reply src/store/index-level.ts:13]: +1 for using less likely characters

[the remark this answers is not in the document]
Comment (dcrousso) [reply src/store/index-level.ts:13]: so i thought about this more and `JSON.stringify` actually wouldn't really work as well cause it could cause issues with searching for something like `gt: 'a'` with a value of `'abcd'` since stringifying it would result in `'a'`, meaning that it would no longer lexicographically compare the same :(

File: src/store/index-level.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/store/index-level.ts]: Sorry for omitting the full context: it's because we really should have `to` in `RangeCriterion` currently only used for dates be exclusive of the value specified. This way a dev can iterate through time ranges without needing to add silly dedup logic. This would be inline with convention in JavaScript array operations (and other languages).

[the remark this answers is not in the document]
Comment (dcrousso) [reply src/store/index-level.ts]: IMO `RangeCriterion` being end range exclusive is kinda a separate issue to whether this very low level structure can support functionality like this.  i would much prefer having the ability to do this at a low level and then a higher level thing choosing not to use it than not even having the ability at any level.

also, why not provide the ability for developers to do `lt` or `lte`?  it seems a bit unfortunate that we would expect the developer to do this work when we already have the ability to do so

File: src/store/index-level.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (dcrousso) [reply src/store/index-level.ts]: i did a bit more playing around with this and i was able to replace the while `KEY_BOUNDARY` by just adding a `findExactMatch` since ultimately that was the only part we were missing

File: src/store/index-level.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/store/index-level.ts]: I am kinda lost. But this one was an optional request 'coz I thought it was low hanging fruit, but feel free to resolve this as out-of-scope for this PR.

**Comment by thehenrytsai** [comment]
@mistermoe (and @dcrousso?!)

> there's only so much one can do with a flat k/v store!

Amen! It definitely is an insane achievement that you were able to take a kv store as far as you did with so few lines of code!! I would've never had the courage to undertake this task so thank you....

File: src/store/index-level.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (dcrousso) [reply src/store/index-level.ts]: so i did even more exploration and i dont think there's a good way around this (at least not in JS)
- i tried leveraging `sublevel` to do the partitioning (in the hopes that it would be smart about encoding things to prevent reserved characters) but apparently `sublevel` restricts keys to be between the one after the reserved character (which defaults to `!`/`\x21`) and `\x1F`, so we'd be unable to use unicode in any value of anything, which is probably not ideal (e.g. querying for a user's name in countries that dont use ASCII)
- i tried using `JSON.stringify` to encode everything but that would affect the lexicographic order of the values, both in what i mentioned above and it transforming unicode (e.g. `\x00` becomes `\\u0000`, which would now compare differently if searching for `\\`)

i think we probably just have to hope that nobody will use any of the reserved characters (which is why i changed them to `\x01` and `\x02`) or other screwy things

**Comment by thehenrytsai** [comment]
@dcrousso, thanks for the amazing effort for eliminating magic chars one by one, in the process making the code even more hardcore for my brain!

If you are done making the core changes, would be great if you can add a few nascent tests to cover all code added (mostly on `throwIfAborted` what not) for completeness. Beyond that I think (more like trust at this point) the code is good to go, I will drool over the intricacies of all the code later!

File: src/store/index-level.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/store/index-level.ts]: Thanks for the due diligence! On board with not using common chars, and you've gone above and beyond at this point, so nothing else to say from my side on this!

**Comment by dcrousso** [comment]
> [...] would be great if you can add a few nascent tests to cover all code added (mostly on `throwIfAborted` what not) for completeness.

@thehenrytsai i dont think it's a good idea to add tests for some of those because they are heavily dependent on the timing of `level` operations

for example, both `put` and `delete` use `Level.prototype.batch`, meaning that there's really only a single `await`-able operation in those methods (not counting creating the `Level` instance as that's already opened by the test harness beforehand).  if i added something like `setTimeout(() => { controller.abort(); }` in the test then it's a race between the `setTimeout` and `level` as to whether the latter can perform the operation(s) quick enough to enqueue something into the event loop (and that doesnt even account for the differences in the event loop system between browser and node).

**Review by thehenrytsai — APPROVED** [review]
